design/ironic-standalone-operator.md

```diff
@@ -0,0 +1,369 @@
+- `nodeSelector` - a selector for nodes to run Ironic on
+- `tlsRef` - a reference to a TLS secret to use for the service
+
+The `networking` sub-structure deserves a separate consideration:
+
+- `apiPort`, `imageServerPort`, `imageServerTLSPort` allow overriding listening
+  ports for the services
+- `bindInterface` - a boolean flag that makes Ironic listen on only the
+  provisioning interface
+- `dhcp` - another nested structure with DHCP parameters (see below)
+- `externalIP` - IP through which nodes deployed over virtual media access
+  Ironic and HTTPD
+- `interface`, `ipAddress`, `macAddresses` - various ways to specify the
+  provisioning interface
+
+Finally, the `dhcp` sub-sub-structure contains the following fields:
+
+- `networkCIDR` - CIDR of the provisioning network (required)
+- `rangeBegin`, `rangeEnd` define the DHCP range (derived from `networkCIDR` if
+  missing)
+- `dnsAddress`, `serveDNS` - two mutually exclusive ways to optionally provide
+  DNS to hosts: either a fixed address or dnsmasq itself
+- `hosts`, `ignore` - fine tuning for specific hosts
+- `gatewayAddress` - IP address of the default gateway (if necessary)
+
+Providing a non-nil `dhcp` value enables dnsmasq.
+
+### Risks and Mitigations
+
+Our reliance on host networking means that it's not trivial to have several
+Ironic installations on the same cluster. Each would need to use different
+ports to avoid conflicts. Even without host networking, having several dnsmasq
+instances on the same network is not going to work without some sort of
+coordination between them.
+
+### Work Items
+
+General enablement:
+
+- Add the operator to the development environment.
+- Add an optional flag either to metal3-dev-env or to BMO e2e tests (TBD) that
+  uses ironic-standalone-operator instead of the Kustomize scripts in BMO.
+- Create and run CI jobs (integration or e2e - depending on the previous work
+  item) on the operator.
+
+HA:
+
+- Implement the boot configuration API in Ironic (dependency).
+- Start generating a private CA for JSON RPC.
+- Enable the HA architecture.
+- Adjust ironic-image to enable updates without wiping the database (see
+  below).
+
+### Dependencies
+
+None for the core operator.
+
+The HA approach will require [boot configuration API][boot config].
+
+### Test Plan
+
+The new operator will become the primary way to install Ironic. As such, it
+will be tested in various CI jobs.
+
+### Upgrade / Downgrade Strategy
+
+By default, the operator will be tightly coupled with the version of Ironic
+(and, eventually, IPA) that it installs. A release of the operator will follow
+each release of ironic-image, and they release branches will match. The `main`
+branch will continue following the latest container image.
+
+After each full reconciliation, the operator will store the version of Ironic
+it has just installed in the `status`. In the future, this will allow to apply
+any logic on upgrade. However, we will try to keep ironic-image self-upgrading
```

Comment:
'However, we will try to keep ironic-image self-upgrading for the sake of users that do not use ironic-standalone-operator.' What do you mean?

Reply:
I mean, if you just use ironic-image, and you upgrade it by replacing the container, we will try to support this case still. Any ideas on how to phrase it better?

Reply:
You mean that for example, if someone was running ironic-standalone-operator v1.0 (with ironic-image v1.0), and just replaced their ironic-image with v1.1, ironic-standalone-operator will continue to work?
If that's what you mean, then 'for the sake of users that do not use ironic-standalone-operator.' can be confusing..

Reply:
Ah, I see the source of confusion. What I wanted to say was: users of ironic-image without the operator will be able to do it still. I'll work on a different way to phrase it.